cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java

```diff
@@ -262,35 +287,169 @@ public TerminationStatus getTerminationStatus() {
   }
 
   /**
-   * Allows more succinct creation of ApiCallback.
+   * Creates a label selector that matches all labels in the meta object.
+   */
+  private String getLabelSelector() {
+    return meta.getLabels().entrySet().stream()
+      .map(e -> String.format("%s=%s", e.getKey(), e.getValue()))
+      .collect(Collectors.joining(","));
+  }
+
+  /**
+   * Restarts the given set of instances. This method currently only supports restart of stateful set.
+   *
+   * @param runnable name of the runnable to restart. This is currently unused
+   * @param instanceIds the set of instance ids to restart
+   * @return a {@link CompletableFuture} that will complete when the delete operation completed
    */
-  private static class CallbackAdapter implements ApiCallback<V1Status> {
-    private final Consumer<ApiException> onFailure;
-    private final Runnable onSuccess;
+  private CompletableFuture<String> doRestartInstances(String runnable, Set<Integer> instanceIds) {
+    CompletableFuture<String> resultFuture = new CompletableFuture<>();
+    CoreV1Api api = new CoreV1Api(apiClient);
+
+    // Only support for stateful set for now
+    if (!V1StatefulSet.class.equals(resourceType)) {
+      resultFuture.completeExceptionally(
+        new UnsupportedOperationException("Instance restart by instance id is only supported for statefulsets"));
+      return resultFuture;
+    }
 
-    CallbackAdapter(Consumer<ApiException> onFailure, Runnable onSuccess) {
-      this.onFailure = onFailure;
-      this.onSuccess = onSuccess;
+    if (instanceIds.size() > 1) {
+      resultFuture.completeExceptionally(
+        new UnsupportedOperationException("Only one instance can be restarted at a time."));
+      return resultFuture;
     }
 
-    @Override
-    public void onFailure(ApiException e, int i, Map<String, List<String>> map) {
-      onFailure.accept(e);
+    int instanceId = instanceIds.iterator().next();
+    V1DeleteOptions deleteOptions = new V1DeleteOptions().preconditions(new V1Preconditions().uid(runnable));
+
+    String podName = String.format("%s-%d", meta.getName(), instanceId);
+    LOG.debug("Deleting pod with name {}", podName);
+
+    try {
+      api.deleteNamespacedPodAsync(podName, kubeNamespace, null, deleteOptions, null, null, null, null,
+                                   createCallbackFutureAdapter(resultFuture, r -> runnable));
```

Comment:
Changes are taken from https://github.com/cdapio/cdap/pull/12522 Except while deleting pod in the statefulset making sure that the uid matches with the one received from poller info so that we dont accidentally delete the pod running some other preview.